api/search.go

```diff
@@ -0,0 +1,374 @@
+		for i, id := range filters.RunIDs {
+			wg.Add(1)
+			go func(i int, id int64) {
+				defer wg.Done()
+
+				var testRun *shared.TestRun
+				testRun, err = sh.simpl.LoadTestRun(id)
+				if err == nil {
+					testRuns[i] = *testRun
+				}
+			}(i, id)
+		}
+		wg.Wait()
+
+		if err != nil {
+			return testRuns, filters, err
+		}
+	}
+
+	return testRuns, filters, nil
+}
+
+func (sh searchHandler) loadSummaries(testRuns []shared.TestRun) ([]summary, error) {
+	var err error
+	summaries := make([]summary, len(testRuns))
+
+	var wg sync.WaitGroup
+	for i, testRun := range testRuns {
+		wg.Add(1)
+
+		go func(i int, testRun shared.TestRun) {
+			defer wg.Done()
+
+			var data []byte
+			s := make(summary)
+			data, loadErr := sh.loadSummary(testRun)
+			if err == nil && loadErr != nil {
+				err = loadErr
+				return
+			}
+			marshalErr := json.Unmarshal(data, &s)
+			if err == nil && marshalErr != nil {
+				err = marshalErr
+				return
+			}
+			summaries[i] = s
+		}(i, testRun)
+	}
+	wg.Wait()
+
+	return summaries, err
+}
+
+func (sh searchHandler) loadSummary(testRun shared.TestRun) ([]byte, error) {
+	mkey := getMemcacheKey(testRun)
+	cached, err := sh.cache.Get(mkey)
+	if cached != nil && err == nil {
+		log.Printf("INFO: Serving summary from cache: %s", mkey)
+		return cached, nil
+	}
+
+	if err != nil {
+		log.Printf("WARNING: Error fetching cache key %s: %v", mkey, err)
+		err = nil
+	}
+
+	url := getResultsURL(testRun, "")
+	log.Printf("INFO: Loading summary from store: %s", url)
+	data, err := sh.store.Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cache summary.
+	go func() {
+		if err := sh.cache.Put(mkey, data); err != nil {
+			log.Printf("WARNING: Failed to write TestRun summary to cache key %s: %v", mkey, err)
+		} else {
+			log.Printf("INFO: Wrote TestRun summary to cache key %s", mkey)
+		}
+	}()
+
+	return data, nil
+}
+
+func prepareResponse(filters shared.SearchFilter, testRuns []shared.TestRun, summaries []summary) SearchResponse {
+	resp := SearchResponse{
+		Runs: testRuns,
+	}
+	// Dedup visited file names via a map of results.
+	resMap := make(map[string]SearchResult)
+	for i, s := range summaries {
+		for filename, passAndTotal := range s {
+			// Exclude filenames that do not match query.
+			if !strings.Contains(filename, filters.Q) {
+				continue
+			}
+
+			if _, ok := resMap[filename]; !ok {
+				resMap[filename] = SearchResult{
+					Name:   filename,
+					Status: make([]SearchRunResult, len(testRuns)),
+				}
+			}
+			resMap[filename].Status[i] = SearchRunResult{
+				Passes: passAndTotal[0],
+				Total:  passAndTotal[1],
+			}
+		}
+	}
+	// Load map into slice and sort it.
+	resp.Results = make([]SearchResult, 0, len(resMap))
+	for _, r := range resMap {
+		resp.Results = append(resp.Results, r)
+	}
+	sort.Sort(byName(resp.Results))
+
+	return resp
+}
+
+func getMemcacheKey(testRun shared.TestRun) string {
+	return "RESULTS_SUMMARY-" + getResultsURL(testRun, "")
```

Comment:
Are we sure the length of keys won't exceed 250 bytes? I didn't have the length of URL in mind when designing the results receiver. And as the URL is constructed from some user (runner) inputs, they might get too long. I suggest using testRun ID or some other metadata that we have control over as the key.

Reply:
I had a similar thought this morning. I'll switch to the run's ID.